src/NzbDrone.Core/DecisionEngine/Specifications/RssSync/HistorySpecification.cs

```diff
@@ -116,6 +124,41 @@ public virtual DownloadSpecDecision IsSatisfiedBy(RemoteEpisode subject, SearchC
                             return DownloadSpecDecision.Reject(DownloadRejectionReason.HistoryUpgradesNotAllowed, "{0} grab event in history and Quality Profile '{1}' does not allow upgrades", rejectionSubject, qualityProfile.Name);
                     }
                 }
+
+                if (episode is { HasFile: true })
+                {
+                    EpisodeHistory availableUsableEpisodeHistoryForCustomFormatScore;
+
+                    var episodeHistoryEventTypeForHistoryComparison = new List<EpisodeHistoryEventType>
```

Comment:
Use a HashSet, that is defined outside of the function so it's not re-created every time.